containers-int-strict/src/Map/Int.hs

```diff
@@ -32,61 +35,62 @@ module Map.Int
        ) where
 
 import Prelude hiding (lookup, null)
+import Data.Coerce (coerce)
 
 import qualified Data.IntMap.Strict as M
 
-newtype Map k v = IM { unIM :: M.IntMap v }
+newtype Map k v = IM (M.IntMap v)
     deriving newtype (Show, Eq)
 
 type Key = (~) Int
 
-empty :: Map k v
-empty = IM M.empty
+empty :: forall k v. Map k v
+empty = coerce @(M.IntMap v) M.empty
 
 singleton :: Key k => k -> v -> Map k v
-singleton k = IM . M.singleton k
+singleton k = coerce . M.singleton k
```

Comment:
Actually, this is not what was desired.. Currently you're using `coerce` for wrapping and unwrapping `newtype`. But you still wrap this newtype, so it still can have performance overhead. The idea is to take advantage of `Coercible` instance for arrow and write functions like this:

```haskell
singleton :: forall k v . Key k => k -> v -> Map k v
singleton = coerce @(k -> v -> M.IntMap v) M.singleton
```

So the idea is to remove all wrapping and unwrapping completely and delegate the implementation to the underlying type.

Reply:
Ah, I see! I wasn't aware of the difference, but that totally makes sense :)